File: python/Q239_sliding_window_maximum.py
#
# @lc app=leetcode id=239 lang=python3
#
# [239] Sliding Window Maximum
#


# @lc code=start
from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        queue: deque[int] = deque()
        result: list[int] = [0] * (len(nums) - k + 1)

        left: int = 0
        for right in range(len(nums)):
            if queue and queue[0] < left:
                queue.popleft()

            while queue and nums[right] >= nums[queue[-1]]:
                queue.pop()

            queue.append(right)

            if right - left + 1 == k:
                result[left] = nums[queue[0]]
                left += 1

        return result


# @lc code=end

File: rust/src/leetcode/Q239_sliding_window_maximum.rs
/*
 * @lc app=leetcode id=239 lang=rust
 *
 * [239] Sliding Window Maximum
 */

// @lc code=start
use std::collections::VecDeque;

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut deque: VecDeque<(usize, i32)> = VecDeque::new();
        let mut result: Vec<i32> = vec![0; nums.len() - k as usize + 1];

        for right in 0..nums.len() {
            let right_value: i32 = nums[right];

            while let Some(&(_, value)) = deque.back() {
                if value < right_value {
                    deque.pop_back();
                } else {
                    break;
                }
            }

            if let Some(&(index, _)) = deque.front() {
                if (index as i32) < (right as i32) - k + 1 {
                    deque.pop_front();
                }
            }

            deque.push_back((right, right_value));

            if right + 1 >= k as usize {
                result[right - k as usize + 1] = deque.front().unwrap().1;
            }
        }

        result
    }
}
// @lc code=end
